palace.pyx

```diff
@@ -1158,6 +1162,125 @@ cdef class SourceGroup:
+
+    def __init__(self, context: Optional[Context]) -> None:
+        if context is None: return
+        self.impl = (<Context> context).impl.create_auxiliary_effect_slot()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type: Optional[Type[BaseException]],
+                 exc_val: Optional[BaseException],
+                 exc_tb: Optional[TracebackType]) -> Optional[bool]:
+        self.destroy()
+
+    def __lt__(self, other: Any) -> bool:
+        if not isinstance(other, AuxiliaryEffectSlot):
+            return NotImplemented
+        return self.impl < (<AuxiliaryEffectSlot> other).impl
+
+    def __le__(self, other: Any) -> bool:
+        if not isinstance(other, AuxiliaryEffectSlot):
+            return NotImplemented
+        return self.impl <= (<AuxiliaryEffectSlot> other).impl
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, AuxiliaryEffectSlot):
+            return NotImplemented
+        return self.impl == (<AuxiliaryEffectSlot> other).impl
+
+    def __ne__(self, other: Any) -> bool:
+        if not isinstance(other, AuxiliaryEffectSlot):
+            return NotImplemented
+        return self.impl != (<AuxiliaryEffectSlot> other).impl
+
+    def __gt__(self, other: Any) -> bool:
+        if not isinstance(other, AuxiliaryEffectSlot):
+            return NotImplemented
+        return self.impl > (<AuxiliaryEffectSlot> other).impl
+
+    def __ge__(self, other: Any) -> bool:
+        if not isinstance(other, AuxiliaryEffectSlot):
+            return NotImplemented
+        return self.impl >= (<AuxiliaryEffectSlot> other).impl
+
+    def __bool__(self) -> bool:
+        return <boolean> self.impl
+
+    def set_gain(self, gain: float) -> None:
+        """If set to true, the reverb effect will automatically
+	apply adjustments to the source's send slot gains based
+	on the effect properties.
+
+        Has no effect when using non-reverb effects. Default is true.
+        """
+        return self.impl.set_gain()
+
+    def send_auto(self, sendauto: bool) -> None:
+        """Update the effect slot with a new effect.
+	The given effect object may be altered or
+	destroyed without affecting the effect slot.
+        """
+        return self.impl.set_send_auto()
+
+    # TODO: apply effect
+
+    def destroy(self) -> None:
+        """Destroys the effect slot, returning it to the system.
+        If the effect slot is currently set on a source send,
+        it will be removed first
+        """
+        return self.impl.destroy()
+
+    @property
+    def source_sends(self) -> List[SourceSend]:
+        """Retrieve each `Source` object and its pairing
+        send this effect slot is set on.
+        """
+        source_sends = []
+        for alure_source_send in self.impl.get_source_sends():
+            source_send = SourceSend(None)
+            source_send.impl = alure_source_send
+            source_sends.append(source_send)
```

Comment:
Instead of list of struct/data class, from the users' point of view, it'd be much better to receive Python data structures, e.g. `List[Tuple[Source, int]]`.